File: Language_Essentials/java/Q4StackQueue.java
// Stack and Queue Operations - TCS NQT
import java.util.LinkedList;
import java.util.Queue;
import java.util.Stack;

public class Q4StackQueue {
    public static void main(String[] args) {
        // --- 1. Stack (LIFO - Last In First Out) ---
        System.out.println("--- Stack Operations ---");
        Stack<Integer> st = new Stack<>();
        
        st.push(10);
        st.push(20);
        st.push(30); // 30 is at the top
        
        System.out.println("Stack Top: " + st.peek());
        
        st.pop(); // Removes 30
        System.out.println("Stack Top after pop: " + st.peek());
        System.out.println("Stack Size: " + st.size());
        
        // --- 2. Queue (FIFO - First In First Out) ---
        // Queue is an interface in Java, implemented by LinkedList
        System.out.println("\n--- Queue Operations ---");
        Queue<String> q = new LinkedList<>();
        
        q.add("First");
        q.add("Second");
        q.add("Third"); // "First" is at the front
        
        System.out.println("Queue Front: " + q.peek());
        
        q.poll(); // Removes "First", returning null if empty
        // Alternatively, q.remove() throws exception if empty
        System.out.println("Queue Front after poll: " + q.peek());
    }
}

File: Language_Essentials/python/q4_stack_queue.py
# Stack and Queue Operations - TCS NQT
from collections import deque

def ds_ops():
    # --- 1. Stack (LIFO - Last In First Out) ---
    print("--- Stack Operations (using list) ---")
    st = []
    
    st.append(10)
    st.append(20)
    st.append(30) # 30 is at the top
    
    print("Stack Top:", st[-1])
    
    st.pop() # Removes 30
    print("Stack Top after pop:", st[-1])
    print("Stack Size:", len(st))
    
    # --- 2. Queue (FIFO - First In First Out) ---
    print("\n--- Queue Operations (using collections.deque) ---")
    # Using a list as a queue is slow for pop(0) -> O(N).
    # Always use deque.
    q = deque()
    
    q.append("First")
    q.append("Second")
    q.append("Third") # "First" is at the front
    
    print("Queue Front:", q[0])
    print("Queue Back:", q[-1])
    
    q.popleft() # Removes "First" in O(1) time
    print("Queue Front after pop:", q[0])

if __name__ == "__main__":
    ds_ops()
